pageserver/src/walreceiver.rs

```diff
@@ -183,12 +195,79 @@ async fn walreceiver_main(
                 }
             }
 
-            ReplicationMessage::PrimaryKeepAlive(_keepalive) => {
-                trace!("received PrimaryKeepAlive");
-                // FIXME: Reply, or the connection will time out
+            ReplicationMessage::PrimaryKeepAlive(keepalive) => {
+                let wal_end = keepalive.wal_end();
+                let timestamp = keepalive.timestamp();
+                let reply_requested: bool = keepalive.reply() != 0;
+
+                trace!(
+                    "received PrimaryKeepAlive(wal_end: {}, timestamp: {} reply: {})",
+                    wal_end,
+                    timestamp,
+                    reply_requested,
+                );
+                if reply_requested {
+                    // TODO: More thought should go into what values are sent here.
+                    let last_lsn = PgLsn::from(pcache.get_last_valid_lsn());
+                    let write_lsn = last_lsn;
+                    let flush_lsn = last_lsn;
+                    let apply_lsn = PgLsn::INVALID;
```

Comment:
I'd like feedback on this part.

Reply:
In a normal streaming replication replica, the values mean:
`write_lsn`: the WAL has been received up to this point. It's not durable stored on disk yet, so the server might still lose it if it crashes.
`flush_lsn`: the WAL has been flushed to disk up to this point. All WAL <= `flush_lsn` should be safe even if the replica crashes
`apply_lsn`: the WAL has been applied up to this point. If you connect to the replica, you should see the effects of all WAL < `apply_lsn`.

Yeah, it's not entirely clear what we should report here, in the page server. `write_lsn` is pretty straightforward. Currently, the page server doesn't write the WAL to disk, so `write_lsn` is not applicable. Pretending that it's the same as `write_lsn` like you have here is probably fine for now. INVALID makes sense for `apply_lsn`. We do apply the WAL in the page server, but we do it differently from a normal standby server, and you can't run queries directly against the page server.

So +1 for this snippet as it is.